diff --git a/pyproject.toml b/pyproject.toml
index 6a11c7a..8045562 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -16,7 +16,7 @@ dependencies = [
     "trimesh",
     "torch>=2",
     "torchvision",
-    "einops",
+    "einops>=0.7.0",
     "huggingface_hub",
     "imageio",
     "numpy<2",
diff --git a/requirements.txt b/requirements.txt
index 878db3f..14744f5 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -2,7 +2,7 @@ pre-commit
 trimesh
 torch>=2
 torchvision
-einops
+einops>=0.7.0
 huggingface_hub
 imageio
 numpy<2
