diff --git a/croco b/croco
--- a/croco
+++ b/croco
@@ -1 +1 @@
-Subproject commit 743ee71a2a9bf57cea6832a9064a70a0597fcfcb
+Subproject commit 743ee71a2a9bf57cea6832a9064a70a0597fcfcb-dirty
diff --git a/dust3r/cloud_opt/init_im_poses.py b/dust3r/cloud_opt/init_im_poses.py
index 7887c5c..1ed2a98 100644
--- a/dust3r/cloud_opt/init_im_poses.py
+++ b/dust3r/cloud_opt/init_im_poses.py
@@ -4,7 +4,7 @@
 # --------------------------------------------------------
 # Initialization functions for global alignment
 # --------------------------------------------------------
-from functools import cache
+from functools import lru_cache
 
 import numpy as np
 import scipy.sparse as sp
@@ -239,7 +239,7 @@ def estimate_focal(pts3d_i, pp=None):
     return float(focal)
 
 
-@cache
+@lru_cache
 def pixel_grid(H, W):
     return np.mgrid[:W, :H].T.astype(np.float32)
 
