diff --git a/mast3r/catmlp_dpt_head.py b/mast3r/catmlp_dpt_head.py
index ac44579..efdc1b1 100644
--- a/mast3r/catmlp_dpt_head.py
+++ b/mast3r/catmlp_dpt_head.py
@@ -11,8 +11,8 @@ import mast3r.utils.path_to_dust3r  # noqa
 from dust3r.heads.postprocess import reg_dense_depth, reg_dense_conf  # noqa
 from dust3r.heads.dpt_head import PixelwiseTaskWithDPT  # noqa
 import dust3r.utils.path_to_croco  # noqa
-from models.blocks import Mlp  # noqa
-from models.dpt_block import Interpolate  # noqa
+from croco.models.blocks import Mlp  # noqa
+from croco.models.dpt_block import Interpolate  # noqa
 
 
 def reg_desc(desc, mode):
